Erasmia Kornelatou, f2821907



# Exploring US Flights Data

In this assignment, you will investigate the [Airline On-Time Performance Data](https://www.transtats.bts.gov/Tables.asp?DB_ID=120&DB_Name=Airline%20On-Time%20Performance%20Data&DB_Short_Name=On-Time#) that we have already met in one of our lectures.

You will use the 2017 data available at https://auebgr-my.sharepoint.com/:u:/g/personal/louridas_aueb_gr/EbeTzde1LddFmqNWBFWWxNcBj5tDR_zvzDHvqqcmnndeEQ?e=Tb1orW.

---

> Panos Louridas, Associate Professor <br />
> Department of Management Science and Technology <br />
> Athens University of Economics and Business <br />
> louridas@aueb.gr

## Question 1

Provide a "misery index" for airports. That is, sort the airports in descending order of the probability that a flight departing from that airport has a delay. Take care of outliers: some airports may have a preposterously low number of flights. We are not interested in them. Your criterion for outliers will be the airports in the lowest 1% percentile in the number of flights.

After you have done that, go around your data again, but this time you will show the average and median delay for each airport. You may sort the results by either.

Finally, enrich your airport misery index by tabulating both the probability, in descending order, that you will experience a delay at a given airport, and the average and median delay that you are likely to experience.

###  Question 1a
Provide a "misery index" for airports. That is, sort the airports in descending order of the probability that a flight departing from that airport has a delay. Take care of outliers: some airports may have a preposterously low number of flights. We are not interested in them. Your criterion for outliers will be the airports in the lowest 1% percentile in the number of flights.

In the first place we should read the flights dataset.

In [1]:
flight_data_2018 = spark\
  .read\
  .option("inferSchema", "true")\
  .option("header", "true")\
  .csv("D:\\Users\\astar\\Desktop\\spark-5-5-20\\flights.csv\\515364771_T_ONTIME_REPORTING.csv")

We will convert `DataFrame` to a table or a view as follows:

In [2]:
flight_data_2018.createOrReplaceTempView("flight_data_2018")

Now let's perform a query in SQL in order to see the structure of our dataset. For a more user-friendly output, we will use the  toPandas() action, which collects all rows and returns a pandas DataFrame.
Here is the sample of the first ten rows of our dataset.

In [3]:
structure_query = spark.sql("""
    SELECT *
    FROM flight_data_2018
    LIMIT 10
    """)
structure_query.toPandas()

FL_DATE TAIL_NUM CARRIER ORIGIN ORIGIN_CITY_NAME DEST    DEST_CITY_NAME  \
0  2018-01-01   N8891A      9E    GFK  Grand Forks, ND  MSP   Minneapolis, MN   
1  2018-01-01   N8891A      9E    MSP  Minneapolis, MN  GFK   Grand Forks, ND   
2  2018-01-01   N293PQ      9E    DTW      Detroit, MI  SAT   San Antonio, TX   
3  2018-01-01   N295PQ      9E    BOS       Boston, MA  CVG    Cincinnati, OH   
4  2018-01-01   N605LR      9E    MSP  Minneapolis, MN  CLT     Charlotte, NC   
5  2018-01-01   N349PQ      9E    SYR     Syracuse, NY  JFK      New York, NY   
6  2018-01-01   N935XJ      9E    GSP        Greer, SC  LGA      New York, NY   
7  2018-01-01   N922XJ      9E    FSD  Sioux Falls, SD  ATL       Atlanta, GA   
8  2018-01-01   N279PQ      9E    DTW      Detroit, MI  IND  Indianapolis, IN   
9  2018-01-01   N137EV      9E    CVG   Cincinnati, OH  MSP   Minneapolis, MN   

   DEP_TIME  DEP_DELAY  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
0      1302       -8.0      1417      -12.0        0.0              None   
1      1124        9.0      1234      -11.0        0.0              None   
2      2022        4.0      2232      -50.0        0.0              None   
3      1349      104.0      1613       94.0        0.0              None   
4       846        1.0      1214      -14.0        0.0              None   
5      1820        5.0      1923      -32.0        0.0              None   
6       656       -4.0       829      -29.0        0.0              None   
7      1011      241.0      1331      221.0        0.0              None   
8      1210       -5.0      1312      -27.0        0.0              None   
9      1052       -8.0      1149      -21.0        0.0              None   

   DIVERTED  CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
0       0.0            NaN            NaN        NaN             NaN   
1       0.0            NaN            NaN        NaN             NaN   
2       0.0            NaN            NaN        NaN             NaN   
3       0.0           79.0            0.0        0.0             0.0   
4       0.0            NaN            NaN        NaN             NaN   
5       0.0            NaN            NaN        NaN             NaN   
6       0.0            NaN            NaN        NaN             NaN   
7       0.0          221.0            0.0        0.0             0.0   
8       0.0            NaN            NaN        NaN             NaN   
9       0.0            NaN            NaN        NaN             NaN   

   LATE_AIRCRAFT_DELAY  _c19  
0                  NaN  None  
1                  NaN  None  
2                  NaN  None  
3                 15.0  None  
4                  NaN  None  
5                  NaN  None  
6                  NaN  None  
7                  0.0  None  
8                  NaN  None  
9                  NaN  None

We will count the flights per Origin.

In [4]:
groupByOrigin = spark.sql("""
    SELECT ORIGIN, count(1) as FlightsPerOrigin
    FROM flight_data_2018
    GROUP BY ORIGIN
    order by count(1)
    """)
groupByOrigin.toPandas()

ORIGIN  FlightsPerOrigin
0      YNG                 2
1      ART                25
2      IFP                45
3      CYS                58
4      AKN                63
5      BKG                71
6      DLG                84
7      GST                84
8      HYA                89
9      ADK               104
10     OWB               111
11     DRT               114
12     PPG               122
13     OGD               126
14     OGS               137
15     HGR               138
16     STC               145
17     MMH               149
18     FLO               170
19     SMX               170
20     EAR               200
21     GUC               230
22     PRC               236
23     WYS               239
24     SPN               257
25     PSM               303
26     VEL               320
27     ELM               333
28     OTH               371
29     CNY               420
..     ...               ...
328    IAD             68700
329    DAL             70995
330    BNA             74263
331    TPA             74526
332    MDW             87095
333    MIA             87880
334    SAN             90699
335    FLL             97513
336    BWI            106739
337    SLC            113054
338    PHL            116946
339    JFK            129684
340    DCA            132517
341    MCO            138296
342    SEA            140006
343    EWR            143875
344    BOS            148191
345    DTW            157518
346    MSP            159375
347    LAS            161113
348    LGA            171093
349    IAH            173806
350    PHX            173962
351    SFO            175849
352    LAX            221486
353    CLT            233317
354    DEN            235989
355    DFW            279298
356    ORD            332953
357    ATL            390046

[358 rows x 2 columns]

At this point we will calculate the lowest 1% percentile in the number of flights which is our criterion for outliers.

In [5]:
threshold = groupByOrigin.toPandas()["FlightsPerOrigin"].quantile(0.01)

In [6]:
threshold

60.85

Based on our results, the number of flights per Origin must be greater than or equal to 60.85 in order not to be consider as an outlier. So, there are 4 outliers - airports and 130 rows will be removed from the dataset:
* 0	YNG	2 
* 1	ART	25
* 2	IFP	45
* 3	CYS	58

In [7]:
new_flight_data_2018 = spark.sql("""
    SELECT *
    FROM flight_data_2018 
    WHERE ORIGIN NOT IN ('YNG', 'ART', 'IFP', 'CYS')
    """)
new_flight_data_2018.createOrReplaceTempView("new_flight_data_2018")

Now we are going to check if the 4 airports have been removed succesfully.

In [8]:
new_groupByOrigin = spark.sql("""
     SELECT ORIGIN, count(1) as FlightsPerOrigin
    FROM new_flight_data_2018
    GROUP BY ORIGIN
    ORDER BY count(1)
    """)
new_groupByOrigin.toPandas()

ORIGIN  FlightsPerOrigin
0      AKN                63
1      BKG                71
2      DLG                84
3      GST                84
4      HYA                89
5      ADK               104
6      OWB               111
7      DRT               114
8      PPG               122
9      OGD               126
10     OGS               137
11     HGR               138
12     STC               145
13     MMH               149
14     FLO               170
15     SMX               170
16     EAR               200
17     GUC               230
18     PRC               236
19     WYS               239
20     SPN               257
21     PSM               303
22     VEL               320
23     ELM               333
24     OTH               371
25     CNY               420
26     MVY               492
27     HHH               535
28     PVU               547
29     LBL               561
..     ...               ...
324    IAD             68700
325    DAL             70995
326    BNA             74263
327    TPA             74526
328    MDW             87095
329    MIA             87880
330    SAN             90699
331    FLL             97513
332    BWI            106739
333    SLC            113054
334    PHL            116946
335    JFK            129684
336    DCA            132517
337    MCO            138296
338    SEA            140006
339    EWR            143875
340    BOS            148191
341    DTW            157518
342    MSP            159375
343    LAS            161113
344    LGA            171093
345    IAH            173806
346    PHX            173962
347    SFO            175849
348    LAX            221486
349    CLT            233317
350    DEN            235989
351    DFW            279298
352    ORD            332953
353    ATL            390046

[354 rows x 2 columns]

We are going to group the number of delayed flights per ORIGIN.

In [9]:
delayed = spark.sql("""
     SELECT ORIGIN, count(1) as DelayedFlightsPerOrigin
    FROM new_flight_data_2018
    WHERE DEP_DELAY > 0
    GROUP BY ORIGIN
    order by count(1)
    """)
delayed.toPandas()

ORIGIN  DelayedFlightsPerOrigin
0      EAR                       17
1      GST                       24
2      AKN                       27
3      DLG                       28
4      DRT                       30
5      BKG                       31
6      ADK                       43
7      OGS                       50
8      LBF                       50
9      HYA                       52
10     STC                       55
11     OWB                       56
12     VEL                       57
13     PPG                       58
14     GUC                       59
15     WYS                       60
16     MMH                       62
17     LWS                       65
18     FLO                       66
19     HGR                       68
20     PRC                       68
21     SMX                       69
22     HHH                       73
23     BFF                       76
24     EKO                       76
25     HIB                       78
26     OGD                       83
27     SPN                       84
28     CNY                       86
29     LBL                       93
..     ...                      ...
324    BNA                    29602
325    STL                    29854
326    MIA                    30191
327    HOU                    31438
328    SAN                    32920
329    SLC                    33103
330    FLL                    37538
331    DAL                    40544
332    PHL                    40660
333    JFK                    41463
334    DCA                    41730
335    MSP                    42913
336    DTW                    43127
337    BWI                    44582
338    IAH                    45808
339    SEA                    47072
340    EWR                    50091
341    MDW                    51131
342    MCO                    53599
343    BOS                    53954
344    LGA                    54408
345    SFO                    61505
346    LAS                    65858
347    PHX                    67482
348    LAX                    77640
349    CLT                    88139
350    DEN                    88915
351    DFW                   100994
352    ORD                   112855
353    ATL                   141182

[354 rows x 2 columns]

We merge the table containing the total flights per Origin with the table of the delayed flights per Origin. 
Then, we calculate the misery index (delayed flights / total flights in descending order) .

In [10]:
new_groupByOrigin.createOrReplaceTempView("new_groupByOrigin")
delayed.createOrReplaceTempView("delayed")
ratio = spark.sql("""
    SELECT new_groupByOrigin.ORIGIN, DelayedFlightsPerOrigin, FlightsPerOrigin, 
            DelayedFlightsPerOrigin/FlightsPerOrigin as misery_index
    FROM new_groupByOrigin, delayed
    WHERE new_groupByOrigin.ORIGIN = delayed.ORIGIN
    ORDER BY misery_index DESC
    """)
ratio.toPandas()

ORIGIN  DelayedFlightsPerOrigin  FlightsPerOrigin  misery_index
0      OGD                       83               126      0.658730
1      SCK                      486               747      0.650602
2      MDW                    51131             87095      0.587072
3      HYA                       52                89      0.584270
4      DAL                    40544             70995      0.571082
5      HOU                    31438             57841      0.543524
6      OWB                       56               111      0.504505
7      HGR                       68               138      0.492754
8      HTS                      423               879      0.481229
9      PPG                       58               122      0.475410
10     BQN                      908              1960      0.463265
11     TTN                     1191              2603      0.457549
12     MVY                      225               492      0.457317
13     STL                    29854             66197      0.450987
14     PSM                      134               303      0.442244
15     OAK                    23154             52676      0.439555
16     LCK                      475              1084      0.438192
17     BKG                       31                71      0.436620
18     BLV                      437              1019      0.428852
19     AKN                       27                63      0.428571
20     BWI                    44582            106739      0.417673
21     MMH                       62               149      0.416107
22     OTH                      154               371      0.415094
23     ADK                       43               104      0.413462
24     BUR                    10950             26583      0.411917
25     LAS                    65858            161113      0.408769
26     SMX                       69               170      0.405882
27     SUX                      478              1189      0.402019
28     BNA                    29602             74263      0.398610
29     ORH                      378               964      0.392116
..     ...                      ...               ...           ...
324    LAR                      112               625      0.179200
325    VEL                       57               320      0.178125
326    SWO                      141               797      0.176913
327    YAK                      127               722      0.175900
328    GCC                      146               833      0.175270
329    BRO                      454              2607      0.174147
330    BRD                      110               637      0.172684
331    PSG                      125               724      0.172652
332    INL                      111               650      0.170769
333    DVL                      105               626      0.167732
334    SIT                      224              1341      0.167040
335    LBL                       93               561      0.165775
336    PIH                      209              1262      0.165610
337    TWF                      209              1271      0.164437
338    HOB                      111               677      0.163959
339    GJT                      611              3750      0.162933
340    FAI                      718              4456      0.161131
341    BRW                      113               727      0.155433
342    RHI                      109               724      0.150552
343    BTM                       96               677      0.141802
344    HHH                       73               535      0.136449
345    HLN                      246              1811      0.135837
346    BFF                       76               573      0.132635
347    SCC                      119               935      0.127273
348    HIB                       78               625      0.124800
349    EKO                       76               647      0.117465
350    CPR                      109               985      0

### Question 1b

After you have done that, go around your data again, but this time you will show the average and median delay for each airport. You may sort the results by either.

In [22]:
from pyspark.sql.functions import *
from pyspark.sql.window import Window
import pyspark.sql.functions as F
median = F.expr('percentile_approx(DEP_DELAY, 0.5)')
mean_median_dat = new_flight_data_2018\
        .groupBy("ORIGIN")\
        .agg(mean("DEP_DELAY").alias('meanOfDepartDelay'), median.alias('medianOfDepartDelay'))\
        .sort(asc("meanOfDepartDelay"))
mean_median_dat.toPandas()

ORIGIN  meanOfDepartDelay  medianOfDepartDelay
0      YAK          -8.720506                -13.0
1      EAR          -7.884422                -13.0
2      SCC          -7.606717                -13.0
3      PSG          -6.861345                -12.0
4      BRW          -6.831933                -11.0
5      CDV          -6.483240                -11.0
6      WRG          -5.875698                -13.0
7      OME          -5.105491                 -9.0
8      BET          -4.588889                -10.0
9      GST          -4.573171                 -8.0
10     JNU          -3.338363                 -9.0
11     LBF          -3.275801                 -9.0
12     OTZ          -3.242075                 -9.0
13     SIT          -3.003782                -10.0
14     ADQ          -2.838188                 -9.0
15     FAI          -2.495587                 -9.0
16     KTN          -2.488584                 -9.0
17     WYS          -2.485356                 -5.0
18     LWS          -2.472067                 -8.0
19     PIH          -2.388580                 -8.0
20     ITO          -1.238088                 -5.0
21     EKO          -0.622084                 -7.0
22     CDC          -0.500805                 -7.0
23     YUM          -0.455669                 -6.0
24     KOA          -0.215824                 -6.0
25     CPR          -0.188391                 -8.0
26     ANC          -0.164208                 -6.0
27     DLG          -0.037037                 -5.0
28     BFF           0.363158                 -8.0
29     TWF           0.615931                 -7.0
..     ...                ...                  ...
324    STC          15.861111                 -2.0
325    DVL          15.945000                 -9.0
326    SHD          16.204651                 -6.0
327    MQT          16.528099                 -5.0
328    USA          17.065373                 -2.0
329    BKG          17.536232                 -1.0
330    HTS          18.103203                  1.0
331    MEI          18.258845                 -5.0
332    CKB          18.586294                 -5.0
333    OGD          19.064000                  5.0
334    LCK          19.771668                 -1.0
335    MVY          19.899582                  0.0
336    SWF          19.902286                 -3.0
337    PGV          20.689988                 -1.0
338    BLV          21.368942                  0.0
339    EGE          21.951767                 -4.0
340    ORH          22.029189                 -3.0
341    SLN          22.279412                 -6.0
342    CMX          22.574530                 -8.0
343    ACK          23.306624                 -3.0
344    PSM          24.225589                  0.0
345    SCK          24.265857                  8.0
346    TTN          25.308151                  0.0
347    LWB          25.709220                 -5.0
348    HGR          26.760870                  0.0
349    OWB          27.635514                  3.0
350    OTH          29.173184                 -2.0
351    MMH          33.503704                 -1.0
352    HYA          34.397727                  3.0
353    PPG          45.385246                  0.0

[354 rows x 3 columns]

### Question 1c
Finally, enrich your airport misery index by tabulating both the probability, in descending order, that you will experience a delay at a given airport, and the average and median delay that you are likely to experience.

In [23]:
ratio.createOrReplaceTempView("ratio")
mean_median_dat.createOrReplaceTempView("mean_median_dat")
mis_mean_med = spark.sql("""
    SELECT ratio.ORIGIN,misery_index, meanOfDepartDelay, medianOfDepartDelay      
    FROM ratio, mean_median_dat
    WHERE ratio.ORIGIN = mean_median_dat.ORIGIN
    ORDER BY ratio.misery_index DESC
    """)
mis_mean_med.toPandas()

ORIGIN  misery_index  meanOfDepartDelay  medianOfDepartDelay
0      OGD      0.658730          19.064000                  5.0
1      SCK      0.650602          24.265857                  8.0
2      MDW      0.587072          14.818530                  4.0
3      HYA      0.584270          34.397727                  3.0
4      DAL      0.571082          14.257589                  3.0
5      HOU      0.543524          12.946117                  2.0
6      OWB      0.504505          27.635514                  3.0
7      HGR      0.492754          26.760870                  0.0
8      HTS      0.481229          18.103203                  1.0
9      PPG      0.475410          45.385246                  0.0
10     BQN      0.463265          15.657570                 -1.0
11     TTN      0.457549          25.308151                  0.0
12     MVY      0.457317          19.899582                  0.0
13     STL      0.450987          10.546727                  0.0
14     PSM      0.442244          24.225589                  0.0
15     OAK      0.439555           8.421102                  0.0
16     LCK      0.438192          19.771668                 -1.0
17     BKG      0.436620          17.536232                 -1.0
18     BLV      0.428852          21.368942                  0.0
19     AKN      0.428571           1.682540                 -4.0
20     BWI      0.417673          11.354090                 -1.0
21     MMH      0.416107          33.503704                 -1.0
22     OTH      0.415094          29.173184                 -2.0
23     ADK      0.413462           3.851485                 -6.0
24     BUR      0.411917           9.518905                 -1.0
25     LAS      0.408769           9.960894                 -1.0
26     SMX      0.405882          12.329412                 -3.0
27     SUX      0.402019          12.415438                 -2.0
28     BNA      0.398610          11.121595                 -1.0
29     ORH      0.392116          22.029189                 -3.0
..     ...           ...                ...                  ...
324    LAR      0.179200           7.851373                 -8.0
325    VEL      0.178125           3.655063                 -8.0
326    SWO      0.176913           2.965340                 -6.0
327    YAK      0.175900          -8.720506                -13.0
328    GCC      0.175270           3.828467                 -8.0
329    BRO      0.174147           6.041568                 -7.0
330    BRD      0.172684           7.841853                 -5.0
331    PSG      0.172652          -6.861345                -12.0
332    INL      0.170769           2.305164                 -9.0
333    DVL      0.167732          15.945000                 -9.0
334    SIT      0.167040          -3.003782                -10.0
335    LBL      0.165775           7.055856                 -7.0
336    PIH      0.165610          -2.388580                 -8.0
337    TWF      0.164437           0.615931                 -7.0
338    HOB      0.163959           2.139466                 -7.0
339    GJT      0.162933           1.798387                 -7.0
340    FAI      0.161131          -2.495587                 -9.0
341    BRW      0.155433          -6.831933                -11.0
342    RHI      0.150552           7.769663                 -7.0
343    BTM      0.141802           0.750000                 -8.0
344    HHH      0.136449           2.687129                 -7.0
345    HLN      0.135837           1.164361                 -5.0
346    BFF      0.132635           0.363158                 -8.0
347    SCC      0.127273          -7.606717                -13.0
348    HIB      0.124800           3.592532                 -5.0
349    EKO      0.117465          -0.622084                 -7.0
350    CPR      0.110660          -0.188391                 -8.0
351    LWS      0.090656          -2.472067                 -8.0
352    LBF      0.087873          -3.275801                 -9.0
353    EAR      0.085000          -7.88

## Question 2

Provide a "misery index" for airlines. That is, do the same thing you did for the airports, but this time we are interested in the airlines that make life difficult for passengers. Sort the airlines in descending order of probability that a flight operated by the airline has a delay. This time we do not care about outliers.


Then, go around your data again, but this time the criterion will be the average and median delay you may expect to have with an airline. Again we do not care about outliers.

And finally, as you would expect,  enrich your airline misery index by tabulating both the probability, in descending order, that you will experience a delay flying with a particular airline, and the average and median delay that you are likely to experience.

Note that because in the 2017 performance data airlines are only indicated by their code, you should also download the carrier lookup table from https://www.transtats.bts.gov/Download_Lookup.asp?Lookup=L_UNIQUE_CARRIERS, so that you can include airline names in your results.

### Question 2a
Provide a "misery index" for airlines. That is, do the same thing you did for the airports, but this time we are interested in the airlines that make life difficult for passengers. Sort the airlines in descending order of probability that a flight operated by the airline has a delay. This time we do not care about outliers.

We will count the flights per carrier.

In [24]:
groupByCarrier = spark.sql("""
     SELECT CARRIER, count(1) as FlightsPerCarrier
    FROM flight_data_2018
    GROUP BY CARRIER
    ORDER BY count(1)
    """)
groupByCarrier.toPandas()

CARRIER  FlightsPerCarrier
0       VX              17670
1       HA              83723
2       G4              96221
3       F9             120035
4       NK             176178
5       EV             202890
6       YV             215138
7       AS             245761
8       9E             245917
9       OH             278457
10      MQ             296001
11      B6             305010
12      YX             316090
13      UA             621565
14      OO             774137
15      AA             916818
16      DL             949283
17      WN            1352552

We are going to group the number of delayed flights per carrier.

In [25]:
delayedCarrier = spark.sql("""
     SELECT CARRIER, count(1) as DelayedFlightsPerCarrier
    FROM flight_data_2018
    WHERE DEP_DELAY > 0
    GROUP BY CARRIER
    order by count(1)
    """)
delayedCarrier.toPandas()

CARRIER  DelayedFlightsPerCarrier
0       VX                      5382
1       HA                     21767
2       G4                     32886
3       NK                     51119
4       F9                     53321
5       EV                     53898
6       YV                     60284
7       9E                     64942
8       AS                     66254
9       YX                     75839
10      MQ                     83546
11      OH                    100247
12      B6                    125011
13      UA                    185249
14      OO                    203193
15      DL                    286165
16      AA                    317751
17      WN                    662363

We merge the table containing the total flights per carrier with the table of the delayed flights per carrier. 
Then, we calculate the misery index (delayed flights / total flights in descending order) .

In [26]:
groupByCarrier.createOrReplaceTempView("groupByCarrier")
delayedCarrier.createOrReplaceTempView("delayedCarrier")
ratio_carrier = spark.sql("""
    SELECT groupByCarrier.CARRIER, DelayedFlightsPerCarrier, FlightsPerCarrier, 
            DelayedFlightsPerCarrier/FlightsPerCarrier as misery_index
    FROM groupByCarrier, delayedCarrier
    WHERE groupByCarrier.CARRIER = delayedCarrier.CARRIER
    ORDER BY misery_index DESC
    """)
ratio_carrier.toPandas()

CARRIER  DelayedFlightsPerCarrier  FlightsPerCarrier  misery_index
0       WN                    662363            1352552      0.489714
1       F9                     53321             120035      0.444212
2       B6                    125011             305010      0.409859
3       OH                    100247             278457      0.360009
4       AA                    317751             916818      0.346580
5       G4                     32886              96221      0.341776
6       VX                      5382              17670      0.304584
7       DL                    286165             949283      0.301454
8       UA                    185249             621565      0.298036
9       NK                     51119             176178      0.290155
10      MQ                     83546             296001      0.282249
11      YV                     60284             215138      0.280211
12      AS                     66254             245761      0.269587
13      EV                     53898             202890      0.265651
14      9E                     64942             245917      0.264081
15      OO                    203193             774137      0.262477
16      HA                     21767              83723      0.259988
17      YX                     75839             316090      0.239929

### Question 2b
Then, go around your data again, but this time the criterion will be the average and median delay you may expect to have with an airline. Again we do not care about outliers.

In [28]:
mean_median_dat2 = flight_data_2018\
        .groupBy("CARRIER")\
        .agg(mean("DEP_DELAY").alias('meanOfDepartDelay'), median.alias('medianOfDepartDelay'))\
        .sort(asc("meanOfDepartDelay"))
mean_median_dat2.toPandas()

CARRIER  meanOfDepartDelay  medianOfDepartDelay
0       HA           0.913099                 -4.0
1       AS           2.481820                 -5.0
2       VX           6.333836                 -4.0
3       YX           7.339655                 -4.0
4       DL           7.461923                 -2.0
5       MQ           7.765102                 -3.0
6       NK           9.868512                 -3.0
7       UA          10.060413                 -3.0
8       AA          10.078218                 -2.0
9       OO          10.112660                 -4.0
10      9E          10.732733                 -4.0
11      WN          11.138968                  0.0
12      YV          11.226541                 -3.0
13      OH          12.038332                 -2.0
14      EV          12.319338                 -4.0
15      G4          12.922575                 -2.0
16      B6          15.869267                 -2.0
17      F9          19.684102                 -1.0

### Question 2c
And finally, as you would expect,  enrich your airline misery index by tabulating both the probability, in descending order, that you will experience a delay flying with a particular airline, and the average and median delay that you are likely to experience.Note that because in the 2017 performance data airlines are only indicated by their code, you should also download the carrier lookup table from https://www.transtats.bts.gov/Download_Lookup.asp?Lookup=L_UNIQUE_CARRIERS, so that you can include airline names in your results.

In the first place we should read the file in the above link and have a look of the structure of it.

In [29]:
carrier_data_2018 = spark\
  .read\
  .option("inferSchema", "true")\
  .option("header", "true")\
  .csv("D:/Users/astar/Desktop\spark-5-5-20/flights.csv/L_UNIQUE_CARRIERS.csv_")

carrier_data_2018.createOrReplaceTempView("carrier_data_2018")
structure_query2 = spark.sql("""
    SELECT *
    FROM carrier_data_2018
    LIMIT 10
    """)
structure_query2.toPandas()

Code                                        Description
0  02Q                                      Titan Airways
1  04Q                                 Tradewind Aviation
2  05Q                                Comlux Aviation, AG
3  06Q                      Master Top Linhas Aereas Ltd.
4  07Q                                Flair Airlines Ltd.
5  09Q  Swift Air, LLC d/b/a Eastern Air Lines d/b/a E...
6  0BQ                                                DCA
7  0CQ                               ACM AIR CHARTER GmbH
8  0FQ               Maine Aviation Aircraft Charter, LLC
9  0GQ       Inter Island Airways, d/b/a Inter Island Air

In [33]:
ratio_carrier.createOrReplaceTempView("ratio_carrier")
mean_median_dat2.createOrReplaceTempView("mean_median_dat2")
mis_mean_med2 = spark.sql("""
    SELECT ratio_carrier.CARRIER,Description,misery_index, meanOfDepartDelay, medianOfDepartDelay      
    FROM ratio_carrier, mean_median_dat2,carrier_data_2018
    WHERE ratio_carrier.CARRIER = mean_median_dat2.CARRIER AND ratio_carrier.CARRIER = carrier_data_2018.Code
    ORDER BY ratio_carrier.misery_index DESC
    """)
mis_mean_med2.toPandas()

CARRIER              Description  misery_index  meanOfDepartDelay  \
0       WN   Southwest Airlines Co.      0.489714          11.138968   
1       F9   Frontier Airlines Inc.      0.444212          19.684102   
2       B6          JetBlue Airways      0.409859          15.869267   
3       OH        PSA Airlines Inc.      0.360009          12.038332   
4       AA   American Airlines Inc.      0.346580          10.078218   
5       G4            Allegiant Air      0.341776          12.922575   
6       VX           Virgin America      0.304584           6.333836   
7       DL     Delta Air Lines Inc.      0.301454           7.461923   
8       UA    United Air Lines Inc.      0.298036          10.060413   
9       NK         Spirit Air Lines      0.290155           9.868512   
10      MQ                Envoy Air      0.282249           7.765102   
11      YV       Mesa Airlines Inc.      0.280211          11.226541   
12      AS     Alaska Airlines Inc.      0.269587           2.481820   
13      EV  ExpressJet Airlines LLC      0.265651          12.319338   
14      9E        Endeavor Air Inc.      0.264081          10.732733   
15      OO    SkyWest Airlines Inc.      0.262477          10.112660   
16      HA   Hawaiian Airlines Inc.      0.259988           0.913099   
17      YX         Republic Airline      0.239929           7.339655   

    medianOfDepartDelay  
0                   0.0  
1                  -1.0  
2                  -2.0  
3                  -2.0  
4                  -2.0  
5                  -2.0  
6                  -4.0  
7                  -2.0  
8                  -3.0  
9                  -3.0  
10                 -3.0  
11                 -3.0  
12                 -5.0  
13                 -4.0  
14                 -4.0  
15                 -4.0  
16                 -4.0  
17                 -4.0